tools/c7n_azure/c7n_azure/resources/storage.py

```diff
@@ -64,156 +65,120 @@ class resource_type(ArmResourceManager.resource_type):
+     from https://www.microsoft.com/en-us/download/details.aspx?id=56519.
 
-    Note that there are firewall rule number limits and that you will likely need to
-    use a regional block to fit within the limit.  The limit for storage accounts is
-    200 rules.
-
-    .. code-block:: yaml
-
-        - type: set-firewall-rules
-              bypass-rules:
-                  - Logging
-                  - Metrics
-              ip-rules:
-                  - 11.12.13.0/16
-                  - ServiceTags.AppService.CentralUS
+     Note that there are firewall rule number limits and that you will likely need to
+     use a regional block to fit within the limit.  The limit for storage accounts is
+     200 rules.
 
+     .. code-block:: yaml
 
-    :example:
+         - type: set-firewall-rules
+               bypass-rules:
+                   - Logging
+                   - Metrics
+               ip-rules:
+                   - 11.12.13.0/16
+                   - ServiceTags.AppService.CentralUS
 
-    Find storage accounts without any firewall rules.
 
-    Configure default-action to ``Deny`` and then allow:
-    - Azure Logging and Metrics services
-    - Two specific IPs
-    - Two subnets
+     :example:
 
-    .. code-block:: yaml
+     Find storage accounts without any firewall rules.
 
-        policies:
-            - name: add-storage-firewall
-              resource: azure.storage
+     Configure default-action to ``Deny`` and then allow:
+     - Azure Logging and Metrics services
+     - Two specific IPs
+     - Two subnets
 
-            filters:
-                - type: value
-                  key: properties.networkAcls.ipRules
-                  value_type: size
-                  op: eq
-                  value: 0
-
-            actions:
-                - type: set-firewall-rules
-                  bypass-rules:
-                      - Logging
-                      - Metrics
-                  ip-rules:
-                      - 11.12.13.0/16
-                      - 21.22.23.24
-                  virtual-network-rules:
-                      - <subnet_resource_id>
-                      - <subnet_resource_id>
+     .. code-block:: yaml
 
-    """
+         policies:
+             - name: add-storage-firewall
+               resource: azure.storage
+
+             filters:
+                 - type: value
+                   key: properties.networkAcls.ipRules
+                   value_type: size
+                   op: eq
+                   value: 0
+
+             actions:
+                 - type: set-firewall-rules
+                   bypass-rules:
+                       - Logging
+                       - Metrics
+                   ip-rules:
+                       - 11.12.13.0/16
+                       - 21.22.23.24
+                   virtual-network-rules:
+                       - <subnet_resource_id>
+                       - <subnet_resource_id>
+
+     """
```

Comment:
From what I can tell an empty action here will disable the current firewall rules also? It might be good to have that example in both because it's not obvious.

Reply:
Oh no for storage if it is empty then no traffic is allowed.  They use a boolean ("default-action") to turn it on and off so it doesn't have the silly behavior of Cosmos DB.

An empty firewall action will just ensure the firewall is enabled in this case.

Reply:
random idea: we can use some non-public IP like 127.0.0.1.. It should restrict all traffic.

Reply:
Well for cosmos we could default to 'append' instead of 'not append' and then empty doesn't disable everything?

Reply:
> random idea: we can use some non-public IP like 127.0.0.1.. It should restrict all traffic.

Yeah but that feels like an exercise for the user? They'd have to do all the same thing if they used arm templates or Azure CLI or Powershell.  If it is going to be confusing it should be confusing the same way as the product is in all our tools, not in some unique Custodian way - we are just an SDK automation tool.

Reply:
I agree with @erwelch and I think `append` needs to default to `true` with the current change.

It was `false` in the original Ivan's PR, but so was a behavior.. If user didn't specify `ip_rules`, we didn't touch them (like PATCH only what specified). Now we do a complete `create_or_update` call.. Right?

https://github.com/cloud-custodian/cloud-custodian/pull/3797/files#diff-822a9bb7efdc6145f17f50538251e85eR56

Reply:
I think the ask is to default append to true globally for all firewall actions? 

Also we could just require the field so they are forced to be aware? 

I'm not sure either really helps with the fact that the Cosmos one works the opposite of all the others.  They still have to call it with append=false and [] to turn it off, and they still have to find some hack to block all traffic.   If the concern is the behavior of an empty cosmos action we should just update the validate method there to block suspect behavior (don't let them use an empty action).  

In terms of the purpose of the action, I think there are two scenarios:

1) Secure all non-compliant firewalls by configuring to corporate standards (append:false)

and 

2) Allow an extra IP, perhaps to let a scanning tool like C7N have access (append:true)

Scenario 1 is probably 99.999999% of users but it is more likely to be destructive in terms of them losing data/config.

Reply:
re: patch - yeah patch only exists for storage so it isn't an option for this, we have to fake it best we can.

Reply:
Thoughts?

```
     Cosmos firewalls are disabled by simply configuring them with empty values.
     To make this action explicit we require you to provide the empty rules array.

     .. code-block:: yaml

        policies:
          - name: cosmos-firewall-clear
            resource: azure.cosmosdb
            filters:
              # The firewall is enabled
              - not:
                - type: value
                  key: properties.ipRangeFilter
                  value: empty
            actions:
              - type: set-firewall-rules
                ip-rules: []
```

Reply:
I think this is fair. If we do differentiate between empty array vs not-specified, it is ok to keep `append` false as default.. 

This way it doesn't wipe your VNET rules if you wanted to change ip-rules

Reply:
Oh no this is something different I think :( 

We will still wipe your vnet rules if you change your IP rules.

I think you are asking for me to check if `self.data['ip-rules'] is None` and if so to not modify this field (and do the same for the other two fields).  

I considered this but I worried it was very complex behavior for a user - now they have to understand patch semantics.

Reply:
Ok it was a pretty significant refactor, but I think I have enhanced to do all of this in the best way possible.

We now do a full `patch` like behavior for both resources.  I emulate this behavior is cosmos bypass/ip-rules in a rather complex way, but it seems to work. 

I also changed the append default.